# Bienvenue !

Avant de commencer cliquez sur Fichier puis Enregistrer une copie.

Ceci est un notebook Jupyter. Il est composé de plusieurs cellules qui sont soit des cellules de code Python (comme la suivante) ou de texte Markdown (celle-ci). Vous pouvez ajouter des cellules avec les boutons +Code et +Texte au dessus.


Pour lancer une cellule vous pouvez
- Dans la barre de menu, cliquer sur Runtime, puis sur Run the focused cell (ou run all...).
- Presser simultanément les touches Ctrl + Entrée.


Pour tous les TP/TD/projet, je vous recommande de faire vos recherches sur Google en anglais. Il y a beaucoup de resources débutant disponibles en français, mais plus vous monterez en compétences et moins vous trouverez d'informations en français. Je vous conseille aussi de privilégier les moteurs de recherche aux IA générationnelles, ou d'utilisez une IA générationnelle comme vous poseriez une question à un professeur : vous pouvez me poser une question sur une notion à la quelle vous avez déjà réfléchi par vous même et comprenez mal, vous ne me demandez pas juste de répondre à une question pour vous.

C'est parti !

In [1]:
%pip install pandas
%pip install matplotlib
%pip install seaborn

   ---------------------------------------- 0.0/11.5 MB ? eta -:--:--
   ---------------------------------------- 0.0/11.5 MB 640.0 kB/s eta 0:00:18
    --------------------------------------- 0.2/11.5 MB 2.7 MB/s eta 0:00:05
   - -------------------------------------- 0.5/11.5 MB 4.3 MB/s eta 0:00:03
   --- ------------------------------------ 0.9/11.5 MB 4.6 MB/s eta 0:00:03
   --- ------------------------------------ 1.1/11.5 MB 5.3 MB/s eta 0:00:02
   ---- ----------------------------------- 1.3/11.5 MB 5.2 MB/s eta 0:00:02
   ----- ---------------------------------- 1.7/11.5 MB 5.4 MB/s eta 0:00:02
   -------- ------------------------------- 2.4/11.5 MB 6.5 MB/s eta 0:00:02
   ---------- ----------------------------- 3.0/11.5 MB 7.4 MB/s eta 0:00:02
   ------------ --------------------------- 3.5/11.5 MB 7.6 MB/s eta 0:00:02
   ------------- -------------------------- 4.0/11.5 MB 8.0 MB/s eta 0:00:01
   --------------- ------------------------ 4.6/11.5 MB 8.5 MB/s eta 0:00:01
   -


[notice] A new release of pip is available: 24.0 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'error'
Note: you may need to restart the kernel to use updated packages.


  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [15 lines of output]
      The 'sklearn' PyPI package is deprecated, use 'scikit-learn'
      rather than 'sklearn' for pip commands.
      
      Here is how to fix this error in the main use cases:
      - use 'pip install scikit-learn' rather than 'pip install sklearn'
      - replace 'sklearn' by 'scikit-learn' in your pip requirements files
        (requirements.txt, setup.py, setup.cfg, Pipfile, etc ...)
      - if the 'sklearn' package is used by one of your dependencies,
        it would be great if you take some time to track which package uses
        'sklearn' instead of 'scikit-learn' and report it to their issue tracker
      - as a last resort, set the environment variable
        SKLEARN_ALLOW_DEPRECATED_SKLEARN_PACKAGE_INSTALL=True to avoid this error
      
      More information is available at
      https://github.com/scikit-learn/sklearn-


[notice] A new release of pip is available: 24.0 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip



   ---------------------------------------- 0.0/294.9 kB ? eta -:--:--
   ---------------------------------------- 0.0/294.9 kB ? eta -:--:--
   - -------------------------------------- 10.2/294.9 kB ? eta -:--:--
   - -------------------------------------- 10.2/294.9 kB ? eta -:--:--
   ---- ---------------------------------- 30.7/294.9 kB 259.2 kB/s eta 0:00:02
   ----- --------------------------------- 41.0/294.9 kB 245.8 kB/s eta 0:00:02
   -------------- ----------------------- 112.6/294.9 kB 544.7 kB/s eta 0:00:01
   ------------------------- ------------ 194.6/294.9 kB 784.3 kB/s eta 0:00:01
   ------------------------------- ------ 245.8/294.9 kB 835.2 kB/s eta 0:00:01
   -------------------------------------- 294.9/294.9 kB 909.5 kB/s eta 0:00:00
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.0 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [2]:
%pip install scikit-learn


[notice] A new release of pip is available: 24.0 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip



     ---------------------------------------- 0.0/60.8 kB ? eta -:--:--
     ---------------------------------------- 0.0/60.8 kB ? eta -:--:--
     ------------------- ------------------ 30.7/60.8 kB 660.6 kB/s eta 0:00:01
     -------------------------------------- 60.8/60.8 kB 815.1 kB/s eta 0:00:00
   ---------------------------------------- 0.0/11.0 MB ? eta -:--:--
   ---------------------------------------- 0.1/11.0 MB 3.3 MB/s eta 0:00:04
   - -------------------------------------- 0.4/11.0 MB 5.6 MB/s eta 0:00:02
   --- ------------------------------------ 1.0/11.0 MB 9.0 MB/s eta 0:00:02
   ----- ---------------------------------- 1.6/11.0 MB 11.4 MB/s eta 0:00:01
   -------- ------------------------------- 2.4/11.0 MB 13.0 MB/s eta 0:00:01
   ----------- ---------------------------- 3.1/11.0 MB 13.3 MB/s eta 0:00:01
   -------------- ------------------------- 3.9/11.0 MB 14.7 MB/s eta 0:00:01
   ------------------ --------------------- 4.9/11.0 MB 15.8 MB/s eta 0:00:01
   -

<span style="background-color:PapayaWhip"> Prenez un moment pour vous familiariser avec l'interface. Cliquez sur les boutons de menus en haut pour voir toutes les possibilités.</span>

In [3]:
# Cette cellule sert à importer des libraires (ou packages) Python
# qui contiennent des fonctions dont nous avons besoin

import numpy as np, pandas as pd
import sklearn
import matplotlib.pyplot as plt
import seaborn as sns

# Manipulation de Python

<span style="background-color:PapayaWhip"> Complétez la cellule suivante pour créer et appeler une fonction permettant de faire la somme d'une liste.</span>

In [4]:
mes_nombres = [3, 8, 2, 5, 6]

def somme_liste(nombres):
    somme = 0
    for n in nombres:
        somme = somme + n

    return somme

print(somme_liste(mes_nombres))

24


<span style="background-color:PapayaWhip"> Créez une fonction permettant de calculer le maximum d'une liste et testez la.</span>

In [5]:
def maximum_liste(nombres):
    maximum = nombres[0]
    for n in nombres:
        if n > maximum:
            maximum = n

    return maximum

print(maximum_liste(mes_nombres))

8



<span style="background-color:PapayaWhip"> En vous servant des 2 fonctions précédantes, créez une fonction prenant en entrée une liste de listes, calcule la somme de chaque liste à l'intérieur et retourne la plus grande somme.</span>

In [6]:
def melimelo(doublelist):
    temp=[]
    for i in range(len(doublelist)):
        temp.append(somme_liste(doublelist[i]))
    return maximum_liste(temp)

print(melimelo([[1, 2, 3], [4, 5, 6], [7, 8, 9]]))

24


<span style="background-color:PapayaWhip"> Utilisez la fonction numpy pour caculer la somme d'une liste en seule ligne. Faites de même pour le maximum.</span>


In [7]:
def somme_liste2(nombres):
    return sum(nombres)

def maximum_liste2(nombres):
    return max(nombres)

print(somme_liste2(mes_nombres))
print(maximum_liste2(mes_nombres))

24
8


# Familiarisation avec pandas

## Données

Les données que nous allons utiliser pour plusieurs TP correspondent à une version modifiée et très réduite du jeu de données [RT-IOT 2022](https://archive.ics.uci.edu/dataset/942/rt-iot2022), qui est un jeu de données de détection d'intrusion sur une infrastructure IOT. Ce jeu de données a été choisi de manière à vous montrer l'intérêt du module non seulement pour les ILIA, mais pour SE et SQR.
Prenez le temps de cliquer sur le lien pour lire le descriptif des données. Regardez également l'article de recherche joint, notamment la figure 2 et le nom des devices IOT inclues.

Ce jeu de données peut être directement importé en utilisant le package du site, la ligne suivante permet de l'installer.

In [8]:
%pip install ucimlrepo

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.0 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


Cliquez sur le lien, copiez collez les instructions permettant d'importer le jeu de données en Python et lancez les dans la cellule suivante.

In [9]:

from ucimlrepo import fetch_ucirepo 
  
# fetch dataset 
rt_iot2022 = fetch_ucirepo(id=942) 
  
# data (as pandas dataframes) 
X = rt_iot2022.data.features 
y = rt_iot2022.data.targets 
  
# metadata 
print(rt_iot2022.metadata) 
  
# variable information 
print(rt_iot2022.variables) 


{'uci_id': 942, 'name': 'RT-IoT2022 ', 'repository_url': 'https://archive.ics.uci.edu/dataset/942/rt-iot2022', 'data_url': 'https://archive.ics.uci.edu/static/public/942/data.csv', 'abstract': 'The RT-IoT2022, a proprietary dataset derived from a real-time IoT infrastructure, is introduced as a comprehensive resource integrating a diverse range of IoT devices and sophisticated network attack methodologies. This dataset encompasses both normal and adversarial network behaviours, providing a general representation of real-world scenarios.\nIncorporating data from IoT devices such as ThingSpeak-LED, Wipro-Bulb, and MQTT-Temp, as well as simulated attack scenarios involving Brute-Force SSH attacks, DDoS attacks using Hping and Slowloris, and Nmap patterns, RT-IoT2022 offers a detailed perspective on the complex nature of network traffic. The bidirectional attributes of network traffic are meticulously captured using the Zeek network monitoring tool and the Flowmeter plugin. Researchers can


Nous allons utiliser pandas, qui permet de gérer des tableaux avec des noms de lignes et des noms de colonnes. Le tableau est de type DataFrame, (parfois raccourci en df). Le nom des lignes se nomme index.
Lorsque le tableau n'a qu'une dimension (une ligne ou une colonne), il est de type Series et non DataFrame.

Parcourez rapidement ce tutoriel pandas : https://pandas.pydata.org/docs/getting_started/intro_tutorials/03_subset_data.html#min-tut-03-subset
. Vous pourrez y revenir par la suite, ce tutoriel vous aidera pour plusieurs des questions suivantes !



<span style="background-color:PapayaWhip"> Comparez les affichages des 3 prochaines cellules. Quelles sont les différences ? Lequel allez-vous utiliser par la suite ?</span>

In [ ]:
X

id.orig_p  id.resp_p proto service  flow_duration  fwd_pkts_tot  \
0           38667       1883   tcp    mqtt      32.011598             9   
1           51143       1883   tcp    mqtt      31.883584             9   
2           44761       1883   tcp    mqtt      32.124053             9   
3           60893       1883   tcp    mqtt      31.961063             9   
4           51087       1883   tcp    mqtt      31.902362             9   
...           ...        ...   ...     ...            ...           ...   
123112      59247      63331   tcp       -       0.000006             1   
123113      59247      64623   tcp       -       0.000007             1   
123114      59247      64680   tcp       -       0.000006             1   
123115      59247      65000   tcp       -       0.000006             1   
123116      59247      65129   tcp       -       0.000006             1   

        bwd_pkts_tot  fwd_data_pkts_tot  bwd_data_pkts_tot  fwd_pkts_per_sec  \
0                  5                  3                  3          0.281148   
1                  5                  3                  3          0.282277   
2                  5                  3                  3          0.280164   
3                  5                  3                  3          0.281593   
4                  5                  3                  3          0.282111   
...              ...                ...                ...               ...   
123112             1                  0                  0     167772.160000   
123113             1                  0                  0     144631.172400   
123114             1                  0                  0     167772.160000   
123115             1                  0                  0     167772.160000   
123116             1                  0                  0     167772.160000   

        ...    active.avg  active.std     idle.min     idle.max     idle.tot  \
0       ...  2.282415e+06         0.0  29729182.96  29729182.96  29729182.96   
1       ...  2.028307e+06         0.0  29855277.06  29855277.06  29855277.06   
2       ...  2.281904e+06         0.0  29842149.02  29842149.02  29842149.02   
3       ...  2.047288e+06         0.0  29913774.97  29913774.97  29913774.97   
4       ...  2.087657e+06         0.0  29814704.90  29814704.90  29814704.90   
...     ...           ...         ...          ...          ...          ...   
123112  ...  5.960464e+00         0.0         0.00         0.00         0.00   
123113  ...  6.914139e+00         0.0         0.00         0.00         0.00   
123114  ...  5.960464e+00         0.0         0.00         0.00         0.00   
123115  ...  5.960464e+00         0.0         0.00         0.00         0.00   
123116  ...  5.960464e+00         0.0         0.00         0.00         0.00   

           idle.avg  idle.std  fwd_init_window_size  bwd_init_window_size  \
0       29729182.96       0.0                 64240                 26847   
1       29855277.06       0.0                 64240                 26847   
2       29842149.02       0.0                 64240                 26847   
3       29913774.97       0.0                 64240                 26847   
4       29814704.90       0.0                 64240                 26847   
...             ...       ...                   ...                   ...   
123112         0.00       0.0                  1024                     0   
123113         0.00       0.0                  1024                     0   
123114         0.00       0.0                  1024                     0   
123115         0.00       0.0                  1024                     0   
123116         0.00       0.0                  1024                     0   

        fwd_last_window_size  
0                        502  
1                        502  
2                        502  
3                        502  
4                        502  
...                      ...  
123112                  1024  
123113      

In [ ]:
print(X)

        id.orig_p  id.resp_p proto service  flow_duration  fwd_pkts_tot  \
0           38667       1883   tcp    mqtt      32.011598             9   
1           51143       1883   tcp    mqtt      31.883584             9   
2           44761       1883   tcp    mqtt      32.124053             9   
3           60893       1883   tcp    mqtt      31.961063             9   
4           51087       1883   tcp    mqtt      31.902362             9   
...           ...        ...   ...     ...            ...           ...   
123112      59247      63331   tcp       -       0.000006             1   
123113      59247      64623   tcp       -       0.000007             1   
123114      59247      64680   tcp       -       0.000006             1   
123115      59247      65000   tcp       -       0.000006             1   
123116      59247      65129   tcp       -       0.000006             1   

        bwd_pkts_tot  fwd_data_pkts_tot  bwd_data_pkts_tot  fwd_pkts_per_sec  \
0                  

In [ ]:
X.head(5)

id.orig_p  id.resp_p proto service  flow_duration  fwd_pkts_tot  \
0      38667       1883   tcp    mqtt      32.011598             9   
1      51143       1883   tcp    mqtt      31.883584             9   
2      44761       1883   tcp    mqtt      32.124053             9   
3      60893       1883   tcp    mqtt      31.961063             9   
4      51087       1883   tcp    mqtt      31.902362             9   

   bwd_pkts_tot  fwd_data_pkts_tot  bwd_data_pkts_tot  fwd_pkts_per_sec  ...  \
0             5                  3                  3          0.281148  ...   
1             5                  3                  3          0.282277  ...   
2             5                  3                  3          0.280164  ...   
3             5                  3                  3          0.281593  ...   
4             5                  3                  3          0.282111  ...   

    active.avg  active.std     idle.min     idle.max     idle.tot  \
0  2282414.913         0.0  29729182.96  29729182.96  29729182.96   
1  2028306.961         0.0  29855277.06  29855277.06  29855277.06   
2  2281903.982         0.0  29842149.02  29842149.02  29842149.02   
3  2047288.179         0.0  29913774.97  29913774.97  29913774.97   
4  2087656.975         0.0  29814704.90  29814704.90  29814704.90   

      idle.avg  idle.std  fwd_init_window_size  bwd_init_window_size  \
0  29729182.96       0.0                 64240                 26847   
1  29855277.06       0.0                 64240                 26847   
2  29842149.02       0.0                 64240                 26847   
3  29913774.97       0.0                 64240                 26847   
4  29814704.90       0.0                 64240                 26847   

   fwd_last_window_size  
0                   502  
1                   502  
2                   502  
3                   502  
4                   502  

[5 rows x 83 columns]

<span style="background-color:PapayaWhip"> Affichez la taille de la dataframe. Combien y a-t-il de lignes ? De colonnes ? </span>

In [10]:
X.shape

(123117, 83)

Il y a 123117 lignes et 83 colonnes dans la dataframe.

<span style="background-color:PapayaWhip"> Affichez le nom des colonnes de la dataframe. </span>

In [11]:
X.columns

Index(['id.orig_p', 'id.resp_p', 'proto', 'service', 'flow_duration',
       'fwd_pkts_tot', 'bwd_pkts_tot', 'fwd_data_pkts_tot',
       'bwd_data_pkts_tot', 'fwd_pkts_per_sec', 'bwd_pkts_per_sec',
       'flow_pkts_per_sec', 'down_up_ratio', 'fwd_header_size_tot',
       'fwd_header_size_min', 'fwd_header_size_max', 'bwd_header_size_tot',
       'bwd_header_size_min', 'bwd_header_size_max', 'flow_FIN_flag_count',
       'flow_SYN_flag_count', 'flow_RST_flag_count', 'fwd_PSH_flag_count',
       'bwd_PSH_flag_count', 'flow_ACK_flag_count', 'fwd_URG_flag_count',
       'bwd_URG_flag_count', 'flow_CWR_flag_count', 'flow_ECE_flag_count',
       'fwd_pkts_payload.min', 'fwd_pkts_payload.max', 'fwd_pkts_payload.tot',
       'fwd_pkts_payload.avg', 'fwd_pkts_payload.std', 'bwd_pkts_payload.min',
       'bwd_pkts_payload.max', 'bwd_pkts_payload.tot', 'bwd_pkts_payload.avg',
       'bwd_pkts_payload.std', 'flow_pkts_payload.min',
       'flow_pkts_payload.max', 'flow_pkts_payload.tot',
       '

<span style="background-color:PapayaWhip"> Complétez les **deux** cellules suivantes </span>

In [12]:
# Il est possible d'accéder aux lignes et colonnes à partir de leur nom
print("Afficher la ligne d'index 82 et la colonne flow_duration :")
display(X.loc[82, 'flow_duration'])

print("Afficher la ligne d'index 82 et toutes les colonnes :")
# Il n'y a qu'une ligne, donc une seule dimension. C'est une Serie non plus une DataFrame
display(X.loc[82])

print("Afficher les lignes d'index 82 et 156 et toutes les colonnes :")
display(X.loc[[82, 156]])

print("Afficher toutes les lignes et la colonne 'flow_duration' :")
display(X['flow_duration'])

print("Afficher toutes les lignes et les colonnes 'flow_duration' et 'service' :")
display(X[['flow_duration', 'service']])

Afficher la ligne d'index 82 et la colonne flow_duration :


31.945694

Afficher la ligne d'index 82 et toutes les colonnes :


id.orig_p                     60057
id.resp_p                      1883
proto                           tcp
service                        mqtt
flow_duration             31.945694
                           ...     
idle.avg                29914595.84
idle.std                        0.0
fwd_init_window_size          64240
bwd_init_window_size          26847
fwd_last_window_size            502
Name: 82, Length: 83, dtype: object

Afficher les lignes d'index 82 et 156 et toutes les colonnes :


id.orig_p  id.resp_p proto service  flow_duration  fwd_pkts_tot  \
82       60057       1883   tcp    mqtt      31.945694             9   
156      52541       1883   tcp    mqtt      32.143276             9   

     bwd_pkts_tot  fwd_data_pkts_tot  bwd_data_pkts_tot  fwd_pkts_per_sec  \
82              5                  3                  3          0.281728   
156             5                  3                  3          0.279996   

     ...   active.avg  active.std     idle.min     idle.max     idle.tot  \
82   ...  2031098.127         0.0  29914595.84  29914595.84  29914595.84   
156  ...  2416835.070         0.0  29726441.14  29726441.14  29726441.14   

        idle.avg  idle.std  fwd_init_window_size  bwd_init_window_size  \
82   29914595.84       0.0                 64240                 26847   
156  29726441.14       0.0                 64240                 26847   

     fwd_last_window_size  
82                    502  
156                   502  

[2 rows x 83 columns]

Afficher toutes les lignes et la colonne 'flow_duration' :


0         32.011598
1         31.883584
2         32.124053
3         31.961063
4         31.902362
            ...    
123112     0.000006
123113     0.000007
123114     0.000006
123115     0.000006
123116     0.000006
Name: flow_duration, Length: 123117, dtype: float64

Afficher toutes les lignes et les colonnes 'flow_duration' et 'service' :


flow_duration service
0           32.011598    mqtt
1           31.883584    mqtt
2           32.124053    mqtt
3           31.961063    mqtt
4           31.902362    mqtt
...               ...     ...
123112       0.000006       -
123113       0.000007       -
123114       0.000006       -
123115       0.000006       -
123116       0.000006       -

[123117 rows x 2 columns]

In [13]:
# Il est aussi possible d'accéder aux lignes et colonnes à partir de leur numéro (attention, ça commence à 0)

print("Afficher la première ligne (toutes les colonnes) :")
display(X.iloc[0]) # Il n'y a qu'une ligne, donc une seule dimension. C'est une Serie et non plus un DataFrame

print("Afficher les ligne d'index 1 à 5 (toutes les colonnes) :")
display(X.iloc[1:6])

print("Afficher les 3 premières colonnes (toutes les lignes) :")
display(X.iloc[:, 0:3])

Afficher la première ligne (toutes les colonnes) :


id.orig_p                     38667
id.resp_p                      1883
proto                           tcp
service                        mqtt
flow_duration             32.011598
                           ...     
idle.avg                29729182.96
idle.std                        0.0
fwd_init_window_size          64240
bwd_init_window_size          26847
fwd_last_window_size            502
Name: 0, Length: 83, dtype: object

Afficher les ligne d'index 1 à 5 (toutes les colonnes) :


id.orig_p  id.resp_p proto service  flow_duration  fwd_pkts_tot  \
1      51143       1883   tcp    mqtt      31.883584             9   
2      44761       1883   tcp    mqtt      32.124053             9   
3      60893       1883   tcp    mqtt      31.961063             9   
4      51087       1883   tcp    mqtt      31.902362             9   
5      48579       1883   tcp    mqtt      31.869686             9   

   bwd_pkts_tot  fwd_data_pkts_tot  bwd_data_pkts_tot  fwd_pkts_per_sec  ...  \
1             5                  3                  3          0.282277  ...   
2             5                  3                  3          0.280164  ...   
3             5                  3                  3          0.281593  ...   
4             5                  3                  3          0.282111  ...   
5             5                  3                  3          0.282400  ...   

    active.avg  active.std     idle.min     idle.max     idle.tot  \
1  2028306.961         0.0  29855277.06  29855277.06  29855277.06   
2  2281903.982         0.0  29842149.02  29842149.02  29842149.02   
3  2047288.179         0.0  29913774.97  29913774.97  29913774.97   
4  2087656.975         0.0  29814704.90  29814704.90  29814704.90   
5  2044137.955         0.0  29825548.17  29825548.17  29825548.17   

      idle.avg  idle.std  fwd_init_window_size  bwd_init_window_size  \
1  29855277.06       0.0                 64240                 26847   
2  29842149.02       0.0                 64240                 26847   
3  29913774.97       0.0                 64240                 26847   
4  29814704.90       0.0                 64240                 26847   
5  29825548.17       0.0                 64240                 26847   

   fwd_last_window_size  
1                   502  
2                   502  
3                   502  
4                   502  
5                   502  

[5 rows x 83 columns]

Afficher les 3 premières colonnes (toutes les lignes) :


id.orig_p  id.resp_p proto
0           38667       1883   tcp
1           51143       1883   tcp
2           44761       1883   tcp
3           60893       1883   tcp
4           51087       1883   tcp
...           ...        ...   ...
123112      59247      63331   tcp
123113      59247      64623   tcp
123114      59247      64680   tcp
123115      59247      65000   tcp
123116      59247      65129   tcp

[123117 rows x 3 columns]

# Nettoyage des données

<span style="background-color:PapayaWhip"> En utilisant https://pandas.pydata.org/docs/reference/api/pandas.Series.value_counts.html
affichez toutes les valeurs présentes dans la variable y (le label) ainsi que le nombre de lignes ayant cette valeur </span>

In [22]:
y.value_counts(dropna=True, ascending=True)


Attack_type               
NMAP_FIN_SCAN                    28
Metasploit_Brute_Force_SSH       37
Wipro_bulb                      253
DDOS_Slowloris                  534
NMAP_TCP_scan                  1002
NMAP_OS_DETECTION              2000
NMAP_XMAS_TREE_SCAN            2010
NMAP_UDP_SCAN                  2590
MQTT_Publish                   4146
ARP_poisioning                 7750
Thing_Speak                    8108
DOS_SYN_Hping                 94659
Name: count, dtype: int64

In [23]:
y.shape

(123117, 1)

En regardant la figure 2 de l'article de recherche (à partir du lien décrivant les données), quelles devices IOT reconnaissez-vous dans la liste des labels ? Laquelle manque ?

- Dans la liste des labels, on aperçoit Thing-speak, Wipro, MQTT.

- Il manque Amazon-Alexa

Créez une liste `normal` contenant les noms corrsepondants à des devices IOT (les autres correspondants à des attaques)
Créez ensuite une variable `is_normal` qui vérifie si chaque valeur dans y est normale ou non. Combien y a-t-il de lignes correspondants à des attaques ?

In [37]:
normal =['MQTT_publish','Thing_speak','Wipro_bulb','Amazon-Alexa']

is_normal = y.isin(normal)
is_normal



Attack_type
0             False
1             False
2             False
3             False
4             False
...             ...
123112        False
123113        False
123114        False
123115        False
123116        False

[123117 rows x 1 columns]

In [39]:
is_normal_true = is_normal[is_normal ].value_counts()
is_normal_true

Attack_type
True           253
Name: count, dtype: int64

In [40]:
is_normal_false = is_normal[~is_normal ].value_counts()
is_normal_false

Attack_type
False          122864
Name: count, dtype: int64

Regardez toutes les colonnes de X. Quelles colonnes ne sont pas numériques ? Supprimez-les de X et vérifiez que toutes les variables sont maintenant numériques.

In [43]:
X_numeric = X.select_dtypes(include=[np.number])
X_numeric

id.orig_p  id.resp_p  flow_duration  fwd_pkts_tot  bwd_pkts_tot  \
0           38667       1883      32.011598             9             5   
1           51143       1883      31.883584             9             5   
2           44761       1883      32.124053             9             5   
3           60893       1883      31.961063             9             5   
4           51087       1883      31.902362             9             5   
...           ...        ...            ...           ...           ...   
123112      59247      63331       0.000006             1             1   
123113      59247      64623       0.000007             1             1   
123114      59247      64680       0.000006             1             1   
123115      59247      65000       0.000006             1             1   
123116      59247      65129       0.000006             1             1   

        fwd_data_pkts_tot  bwd_data_pkts_tot  fwd_pkts_per_sec  \
0                       3                  3          0.281148   
1                       3                  3          0.282277   
2                       3                  3          0.280164   
3                       3                  3          0.281593   
4                       3                  3          0.282111   
...                   ...                ...               ...   
123112                  0                  0     167772.160000   
123113                  0                  0     144631.172400   
123114                  0                  0     167772.160000   
123115                  0                  0     167772.160000   
123116                  0                  0     167772.160000   

        bwd_pkts_per_sec  flow_pkts_per_sec  ...    active.avg  active.std  \
0               0.156193           0.437341  ...  2.282415e+06         0.0   
1               0.156821           0.439097  ...  2.028307e+06         0.0   
2               0.155647           0.435811  ...  2.281904e+06         0.0   
3               0.156440           0.438033  ...  2.047288e+06         0.0   
4               0.156728           0.438839  ...  2.087657e+06         0.0   
...                  ...                ...  ...           ...         ...   
123112     167772.160000      335544.320000  ...  5.960464e+00         0.0   
123113     144631.172400      289262.344800  ...  6.914139e+00         0.0   
123114     167772.160000      335544.320000  ...  5.960464e+00         0.0   
123115     167772.160000      335544.320000  ...  5.960464e+00         0.0   
123116     167772.160000      335544.320000  ...  5.960464e+00         0.0   

           idle.min     idle.max     idle.tot     idle.avg  idle.std  \
0       29729182.96  29729182.96  29729182.96  29729182.96       0.0   
1       29855277.06  29855277.06  29855277.06  29855277.06       0.0   
2       29842149.02  29842149.02  29842149.02  29842149.02       0.0   
3       29913774.97  29913774.97  29913774.97  29913774.97       0.0   
4       29814704.90  29814704.90  29814704.90  29814704.90       0.0   
...             ...          ...          ...          ...       ...   
123112         0.00         0.00         0.00         0.00       0.0   
123113         0.00         0.00         0.00         0.00       0.0   
123114         0.00         0.00         0.00         0.00       0.0   
123115         0.00         0.00         0.00         0.00       0.0   
123116         0.00         0.00         0.00         0.00       0.0   

        fwd_init_window_size  bwd_init_window_size  fwd_last_window_size  
0                      64240                 26847                   502  
1                      64240                 26847                   502  
2                      64240                 26847                   502  
3                      64240                 26847                   502  
4                      64240                 26847                   502  
...                      ...                   ...            

# Bonus

Affichez les valeurs moyennes, l'écart type, le min et le max de chaque variable. Affichez l'histogramme de variables vous semblant intéressantes, ainsi que le scatter plot d'un couple de variables pour voir leur relation

In [45]:
X_numeric.describe()
    

id.orig_p      id.resp_p  flow_duration   fwd_pkts_tot  \
count  123117.000000  123117.000000  123117.000000  123117.000000   
mean    34639.258738    1014.305092       3.809566       2.268826   
std     19070.620354    5256.371994     130.005408      22.336565   
min         0.000000       0.000000       0.000000       0.000000   
25%     17702.000000      21.000000       0.000001       1.000000   
50%     37221.000000      21.000000       0.000004       1.000000   
75%     50971.000000      21.000000       0.000005       1.000000   
max     65535.000000   65389.000000   21728.335580    4345.000000   

        bwd_pkts_tot  fwd_data_pkts_tot  bwd_data_pkts_tot  fwd_pkts_per_sec  \
count  123117.000000      123117.000000      123117.000000      1.231170e+05   
mean        1.909509           1.471218           0.820260      3.518063e+05   
std        33.018311          19.635196          32.293948      3.707645e+05   
min         0.000000           0.000000           0.000000      0.000000e+00   
25%         1.000000           1.000000           0.000000      7.454354e+01   
50%         1.000000           1.000000           0.000000      2.467238e+05   
75%         1.000000           1.000000           0.000000      5.242880e+05   
max     10112.000000        4345.000000       10105.000000      1.048576e+06   

       bwd_pkts_per_sec  flow_pkts_per_sec  ...    active.avg    active.std  \
count      1.231170e+05       1.231170e+05  ...  1.231170e+05  1.231170e+05   
mean       3.517620e+05       7.035683e+05  ...  1.481354e+05  2.353599e+04   
std        3.708015e+05       7.415634e+05  ...  1.613007e+06  1.477935e+06   
min        0.000000e+00       0.000000e+00  ...  0.000000e+00  0.000000e+00   
25%        7.288927e+01       1.490871e+02  ...  9.536740e-01  0.000000e+00   
50%        2.467238e+05       4.934475e+05  ...  4.053116e+00  0.000000e+00   
75%        5.242880e+05       1.048576e+06  ...  5.006790e+00  0.000000e+00   
max        1.048576e+06       2.097152e+06  ...  4.374931e+08  4.774862e+08   

           idle.min      idle.max      idle.tot      idle.avg      idle.std  \
count  1.231170e+05  1.231170e+05  1.231170e+05  1.231170e+05  1.231170e+05   
mean   1.616655e+06  1.701956e+06  3.517644e+06  1.664985e+06  4.550183e+04   
std    8.809396e+06  9.252337e+06  1.229508e+08  9.007064e+06  1.091361e+06   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
50%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
75%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
max    3.000000e+08  3.000000e+08  2.096777e+10  3.000000e+08  1.208029e+08   

       fwd_init_window_size  bwd_init_window_size  fwd_last_window_size  
count         123117.000000         123117.000000         123117.000000  
mean            6118.905123           2739.776018            751.647514  
std            18716.313861          10018.848534           6310.183843  
min                0.000000              0.000000              0.000000  
25%               64.000000              0.000000             64.000000  
50%               64.000000              0.000000             64.000000  
75%               64.000000              0.000000             64.000000  
max            65535.000000          65535.000000          65535.000000  

[8 rows x 81 columns]